TINY-7318 (#2021)
tinymce/tinymce-docs: rtc/configuration.md

})
```

## Recommended configuration

For the best user experience, {{site.companyname}} recommends including these configuration settings:

* [`rtc_server_disconnected`](#rtc_server_disconnected)
* [`rtc_user_details_provider`](#rtc_user_details_provider)

### `rtc_server_disconnected`

If the RTC session fails to connect, or is disconnected due to an error, the user will be blocked from editing (using [setProgressState]({{site.baseurl}}/api/tinymce/tinymce.editor/#setprogressstate)) along with an error notification:

> ![RTC disconnected error example]({{site.baseurl}}/images/rtc-error-example.png "RTC disconnected error example")

Reviewer: I was able to slim this image down by 55% with ImageOptim. Probably worth doing

![Screen Shot 2021-07-07 at 1 37 18 pm](https://user-images.githubusercontent.com/2288242/124696448-746e3e80-df28-11eb-94db-a5fd35823aa6.png)

Author: I always forget to do that - but to be fair I wasn't even sure there should _be_ a screenshot. I was hoping for a doc or PM review before this progresses.

Author: I ran tinypng on the whole image folder - it shrinks the RTC error png a lot more than ImageOptim (final size 9404 bytes) - and saved 2,534,787 bytes. I'll make a separate PR for that, though.

> The server disconnected callback can be used to provide an integration response to this condition. The function is provided both a reason for the disconnect and a suggested HTML error message to display to the user.

Reviewer: "The server disconnected callback can be used to provide an integration response to this condition." This reads a bit fluffy, might not even need it?

Author: I'm describing the fields. I'll see if I can move the wording down to the field description and it still makes sense.

> > **Caution**: It is critical to at least handle the `client_update_required` reason. This indicates the plugin is out of date compared to other users on the session. The behaviour in this scenario depends on the configuration:

Reviewer: Maybe just for clarity?
```suggestion
> **Caution**: It is critical to at least handle the `client_update_required` reason. This indicates the RTC plugin is out of date compared to other users on the session. The behavior in this scenario depends on the configuration:
```

**Type:** `Function`

**Required:** no

#### Input fields for `rtc_server_disconnected`

| Field | Type | Description |
|-------|:----:|-------------|
| `reason` | `string` | One of the reasons described below. |
| `message` | `string` | A translated description suitable for displaying to users. This string might contain HTML, and in some cases is the same string displayed in the editor notification. |

#### Reasons for disconnection

The `reason` field will have one of the following values. 

| Value |  Description |
|-------|-------------|
| `client_update_required` | This error indicates the RTC plugin is out of date and cannot connect to an active session for the supplied `rtc_document_id`. This can happen on startup, but is more common at runtime during editor upgrades. The suggested message recommends the user reload the page. |
| `encryption` | Indicates a failure on startup either in the cryptography process or `rtc_encryption_provider`. This usually means there is an error in the editor configuration. |
| `jwt` | Indicates a problem with `rtc_token_provider`. Either the provider returned a rejected promise or the token was invalid. |

> | `content` | Indicates a problem with `rtc_initial_content_provider`. The only time this generally happens is when the provider returns a rejected promise. |

Reviewer: "The only" conflicts with "generally"
```suggestion
| `content` | Indicates a problem with `rtc_initial_content_provider`. This generally happens when the provider returns a rejected promise. |
```

Author: Weird wording, I agree. I didn't want to make it definitive in case it changes (there is actually only one reason in the code right now) but it did make it to vague. I'll change it completely.

> | `general` | A generic error for reasons that do not yet have a category. Details will be printed to the browser console. |

Author: This seems unfinished? I guess we can tweak the formatting later

#### Example of handling server disconnection

```js
tinymce.init({
  selector: 'textarea', // change this value according to your HTML
  plugins: 'rtc',
  rtc_server_disconnected: ({reason, message}) => {
    console.log('Disconnected', reason, message);

>     // perform some action in response to the RTC session disconnecting

Author: can I log a ticket for that? 😂

>     // perform some action in response to the RTC session disconnecting

Reviewer 2: I hope to fix this in the coming weeks. if not, yes?

### `rtc_snapshot`

Real-time collaboration integrations regularly store the content, eliminating the need for a save button. The {{site.productname}} RTC plugin provides a version number to assist with storing the regular content snapshots. These snapshots are not stored by the {{site.cloudname}} and must be handled by the integrator.

For any given document ID, the server guarantees the version number will only increase. It can be safely used for conflict resolution. For each document ID and version combination the snapshot content is guaranteed to be identical.

The snapshot callback will be executed at regular intervals with access to the serialized editor content. The content is retrieved through a `getContent` function to reduce CPU load if the callback decides to not use the editor content.

**Type:** `Function`

**Required:** no

#### Input fields for `rtc_snapshot`

| Field | Type | Description |
|-------|:----:|-------------|
| `documentId` | `string` | The document ID from the [`rtc_document_id`](#rtc_document_id) option
| `version` | `integer` | An increasing version number, specific to the current document ID, between 0 and 2147483648 (2<sup>31</sup>). |

> | `getContent()` | `string` | Function to execute to get the content for that particular version. |

Reviewer: This looks a little weird, should it be

```suggestion
| `getContent` | `Function` | Function to execute to get the content for that particular version. |
```

Author: One of those things that snuck through a previous review and I've just moved it. That probably makes sense.

Only one `rtc_client_connected` event will be fired per client connection. Connecting to a session with multiple existing clients will fire separate `rtc_client_connected` events for each existing client.

**Type:** `Function`

**Required:** no

#### Input fields for `rtc_client_connected`

> To help with generating a user interface for connected users, an object with multiple fields is provided.

Reviewer 2: ```suggestion
To help with generating a user interface for connected users, an object for user details is provided with the following fields:

- [User ID (`userId`)](#userid)
- [User Details (`userDetails`)](#userdetails)
- [Client ID (`clientId`)](#clientid)
- [Caret Number (`caretNumber`)](#caretnumber)
- [Client information (`clientInfo`)](#clientinformation)
```
or something like that.

#### User ID

| Field | Type |
|-------|:----:|
| `userId` | `string` |

This is the user's unique ID (the `sub` field from their [JWT](({{site.baseurl}}/rtc/jwt-authentication/)), which is also used for [`rtc_user_details_provider`](#rtc_user_details_provider)). Multiple connection events will be received with the same user ID if a user opens multiple sessions (for example on desktop and mobile).

#### User Details

| Field | Type |
|-------|:----:|
| `userDetails` | `object` |

This is a copy of the object returned by [`rtc_user_details_provider`](#rtc_user_details_provider). RTC only uses the `fullName` field, but the entire object will be cloned and passed to `rtc_client_connected`.

If a user details provider is not configured, this will be an empty object.

#### Client ID

| Field | Type |
|-------|:----:|
| `clientId` | `string` |

This is a unique identifier, generated by the RTC protocol, that can be used to differentiate between the same user connecting multiple times.

#### Caret Number

| Field | Type |
|-------|:----:|
| `caretNumber` | `integer` |

> This will be a number between 1 and 8, corresponding to one of [the 8 colours defined in TinyMCE CSS](https://github.com/tinymce/tinymce/blob/master/modules/oxide/src/less/theme/content/rtc/rtc.less#L1-L8). TinyMCE supports 8 distinct caret colors. If more than 8 clients connect to a session, the numbers will be reused.

Reviewer: ```suggestion
This will be a number between 1 and 8, corresponding to one of [the 8 colors defined in TinyMCE CSS](https://github.com/tinymce/tinymce/blob/master/modules/oxide/src/less/theme/content/rtc/rtc.less#L1-L8). TinyMCE supports 8 distinct caret colors. If more than 8 clients connect to a session, the numbers will be reused.
```

> A custom skin is required to change these colours, and no more than 8 are supported in this release. For information on creating a custom skin, see: [Customizing the Editor UI]({{site.baseurl}}/general-configuration-guide/customize-ui/).

Reviewer: ```suggestion
A custom skin is required to change these colors, and no more than 8 are supported in this release. For information on creating a custom skin, see: [Customizing the Editor UI]({{site.baseurl}}/general-configuration-guide/customize-ui/).
```

#### Client information

| Field | Type |
|-------|:----:|
| `clientInfo` | `object` |

This is a copy of the [`rtc_client_info`](#rtc_client_info) data from the remote user's editor configuration.

If none was configured, this will be an empty object.

> **Caution**: {{site.productname}} cannot guarantee the accuracy of data which comes from a remote object. {{site.companyname}} recommends only using the client information data for status flags. To obtain authentic client information, use the [`rtc_user_details_provider`](#rtc_user_details_provider) data returned through the `userDetails` field.

#### Example of client connected function

> See below.

Reviewer: This is a bit dangerous if content moves and changes

Author: I don't like many things about this section, but at least it's easier to follow now. @tylerkelly13 has said he'd look at ways to improve it 🤔

### Example using client connect and disconnect events with custom user details

While all fields are provided to both `rtc_client_connected` and `rtc_client_disconnected` functions, this example shows only relevant fields need to be used.

```js
const connectedUsers = {}
tinymce.init({
  selector: 'textarea', // change this value according to your HTML
  plugins: 'rtc',
  rtc_user_details_provider: ({userId}) => Promise.resolve({ fullName: "John Doe", jobTitle: "Engineer" }),

>   rtc_client_connected: ({userDetails: {fullName, jobTitle}, userId, caretNumber, clientId, clientInfo}) => {

Reviewer: I realise this part is unchanged, but with the amount of info in here now it might be worth only doing one level of de-structuring?

Author: Yes, this is just me sneaking in some developer education with code golf. I also think it looks better than later doing 
```
console.log(`${userDetails.jobTitle} ${userDetails.fullName}....
```

    connectedUsers[clientId] = {
      caretNumber,
      clientInfo,
      userDetails,
      userId
    }
    console.log(`${jobTitle} ${fullName} connected with caret number ${caretNumber}`, clientInfo)
  }
  rtc_client_disconnected: ({clientId, userDetails: {fullName, jobTitle}}) => {
    delete connectedUsers[clientId]
    console.log(`${jobTitle} ${fullName} disconnected`)
  }
})

> ```

Author: I can try? 🤔

> ```

Author: It's going to be weird having both connected and disconnected in both examples but I don't want to separate them. I'll do it and see how it looks.
